File: app/src/main/java/social/entourage/android/location/LocationFragment.java
package social.entourage.android.location;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;
import androidx.core.content.PermissionChecker;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlaceSelectionListener;
import com.google.android.gms.location.places.ui.SupportPlaceAutocompleteFragment;
import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMapOptions;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MapStyleOptions;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.List;
import java.util.Locale;
import java.util.TimerTask;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.EntourageLocation;
import social.entourage.android.EntourageSecuredActivity;
import social.entourage.android.R;
import social.entourage.android.api.tape.Events;
import social.entourage.android.base.EntourageDialogFragment;
import social.entourage.android.tools.BusProvider;
import timber.log.Timber;

/**
 * Fragment to choose the location of an entourage
 * Activities that contain this fragment must implement the
 * {@link LocationFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link LocationFragment#newInstance} factory method to
 * create an instance of this fragment.
 */

public class LocationFragment extends EntourageDialogFragment {

    // ----------------------------------
    // Constants
    // ----------------------------------

    public static final String TAG = "social.entourage.android.entourage.location";

    private static final String KEY_ENTOURAGE_LOCATION = "social.entourage.android.KEY_ENTOURAGE_LOCATION";
    private static final String KEY_ENTOURAGE_ADDRESS = "social.entourage.android.KEY_ENTOURAGE_ADDRESS";
    private static final String KEY_USE_GOOGLE_PLACES_ONLY = "social.entourage.android.KEY_USE_GOOGLE_PLACES_ONLY";

    private static final float LOCATION_MOVE_DELTA = 50; //meters

    private static final float LOCATION_SEARCH_RADIUS = 0.18f; // 20 kilometers in lat/long degrees

    private static final int PERMISSIONS_REQUEST_LOCATION = 1;

    private static final long SEARCH_DELAY = 1000L;

    // ----------------------------------
    // Attributes
    // ----------------------------------

    private LatLng originalLocation;
    private String originalAddress;
    private boolean useGooglePlacesOnly = false;

    private LatLng location;
    private Place selectedPlace;

    private OnFragmentInteractionListener mListener;

    View toReturn;

    @BindView(R.id.entourage_location_address)
    TextView addressTextView;

    SupportMapFragment mapFragment;
    GoogleMap map;

    SupportPlaceAutocompleteFragment autocompleteFragment = null;

    private GeocoderAddressTask geocoderAddressTask = null;

    Handler handler;
    TimerTask timerTask;
    boolean fromPlaceSelected = false;

    private Marker pin;

    // ----------------------------------
    // Lifecycle
    // ----------------------------------

    public LocationFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param location The initial location.
     * @param address The initial address as string
     * @param listener The listener
     * @return A new instance of fragment LocationFragment.
     */

    public static LocationFragment newInstance(LatLng location, String address, OnFragmentInteractionListener listener) {
        LocationFragment fragment = new LocationFragment();
        fragment.mListener = listener;
        Bundle args = new Bundle();
        args.putParcelable(KEY_ENTOURAGE_LOCATION, location);
        args.putString(KEY_ENTOURAGE_ADDRESS, address);
        fragment.setArguments(args);
        return fragment;
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param location The initial location.
     * @param address The initial address as string
     * @param useGooglePlacesOnly true if the fragment will use only Google Places to fetch an address
     * @param listener The listener
     * @return A new instance of fragment LocationFragment.
     */

    public static LocationFragment newInstance(LatLng location, String address, boolean useGooglePlacesOnly, OnFragmentInteractionListener listener) {
        LocationFragment fragment = new LocationFragment();
        fragment.mListener = listener;
        Bundle args = new Bundle();
        args.putParcelable(KEY_ENTOURAGE_LOCATION, location);
        args.putString(KEY_ENTOURAGE_ADDRESS, address);
        args.putBoolean(KEY_USE_GOOGLE_PLACES_ONLY, useGooglePlacesOnly);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            originalLocation = getArguments().getParcelable(KEY_ENTOURAGE_LOCATION);
            if (originalLocation != null) {
                location = new LatLng(originalLocation.latitude, originalLocation.longitude);
            }
            originalAddress = getArguments().getString(KEY_ENTOURAGE_ADDRESS);
            useGooglePlacesOnly = getArguments().getBoolean(KEY_USE_GOOGLE_PLACES_ONLY, false);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        // Inflate the layout for this fragment
        toReturn = inflater.inflate(R.layout.fragment_entourage_location, container, false);
        ButterKnife.bind(this, toReturn);

        toReturn.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                toReturn.getViewTreeObserver().removeOnGlobalLayoutListener(this);

                int h = toReturn.getHeight();
                if (getDialog() != null) {
                    Window window = getDialog().getWindow();
                    if (window != null) {
                        WindowManager.LayoutParams layoutParams = window.getAttributes();
                        layoutParams.height = h;
                        window.setAttributes(layoutParams);
                    }
                }
            }
        });

        return toReturn;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initializeView();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
        cancelTimer();
    }

    @Override
    public void onRequestPermissionsResult(final int requestCode, @NonNull final String[] permissions, @NonNull final int[] grantResults) {
        if (requestCode == PERMISSIONS_REQUEST_LOCATION) {
            EntourageSecuredActivity activity = (EntourageSecuredActivity) getActivity();
            if (activity != null && map != null) {
                for (int index = 0; index < permissions.length; index++) {
                    if (permissions[index].equalsIgnoreCase(activity.getUserLocationAccess()) && grantResults[index] == PackageManager.PERMISSION_GRANTED) {
                        BusProvider.getInstance().post(new Events.OnLocationPermissionGranted(true));
                        try {
                            map.setMyLocationEnabled(true);
                        } catch (SecurityException ex) {
                            Timber.e(ex);
                        }
                    } else {
                        BusProvider.getInstance().post(new Events.OnLocationPermissionGranted(false));
                    }
                }
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    // ----------------------------------
    // INTERFACE CALLBACKS
    // ----------------------------------

    @OnClick(R.id.title_close_button)
    protected void onBackClicked() {
        dismiss();
    }

    @OnClick(R.id.title_action_button)
    protected void onValidateClicked() {
        if (mListener != null) {
            mListener.onEntourageLocationChosen(location, addressTextView.getText().toString(), selectedPlace);
        }
        cancelTimer();
        dismiss();
    }

    @OnClick(R.id.entourage_location_current_position)
    protected void onCurrentLocationClicked() {
        if (map != null) {
            EntourageSecuredActivity activity = (EntourageSecuredActivity) getActivity();
            if (activity != null) {
                if (activity.isGeolocationGranted()) {
                    Location currentLocation = EntourageLocation.getInstance().getCurrentLocation();
                    if (currentLocation != null) {
                        CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLng(new LatLng(currentLocation.getLatitude(), currentLocation.getLongitude()));
                        map.moveCamera(cameraUpdate);
                    }
                } else {
                    requestPermissions(new String[]{activity.getUserLocationAccess()}, PERMISSIONS_REQUEST_LOCATION);
                }
            }
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------

    private void initializeView() {
        // Initialize map
        initializeMap();
        initializePlaces();
        initializeTimer();

        // Initialize address
        if (originalAddress != null) {
            setAddress(originalAddress, false);
        }
    }

    private void initializeMap() {
        if (mapFragment == null) {
            GoogleMapOptions googleMapOptions = new GoogleMapOptions();
            googleMapOptions.zOrderOnTop(true);
            mapFragment = SupportMapFragment.newInstance(googleMapOptions);
        }
        FragmentManager fragmentManager = getChildFragmentManager();
        fragmentManager.beginTransaction().replace(R.id.entourage_location_map_layout, mapFragment).commit();

        mapFragment.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(final GoogleMap googleMap) {
                if (getActivity() != null) {
                    if ((PermissionChecker.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)
                            || (PermissionChecker.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED)) {
                        googleMap.setMyLocationEnabled(true);
                    }
                }
                googleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(
                        getActivity(), R.raw.map_styles_json));
                googleMap.getUiSettings().setMyLocationButtonEnabled(false);
                googleMap.getUiSettings().setMapToolbarEnabled(false);

                if (originalLocation != null) {
                    CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLngZoom(originalLocation, 15);
                    googleMap.moveCamera(cameraUpdate);

                    MarkerOptions markerOptions = new MarkerOptions()
                            .position(originalLocation)
                            .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_pin_orange));
                    pin = googleMap.addMarker(markerOptions);
                } else {
                    MarkerOptions markerOptions = new MarkerOptions()
                            .position(googleMap.getCameraPosition().target)
                            .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_pin_orange));
                    pin = googleMap.addMarker(markerOptions);
                }

                googleMap.setOnCameraMoveStartedListener(new GoogleMap.OnCameraMoveStartedListener() {
                    @Override
                    public void onCameraMoveStarted(final int i) {
                        if (i == GoogleMap.OnCameraMoveStartedListener.REASON_GESTURE) {
                            cancelTimer();
                        }
                    }
                });

                googleMap.setOnCameraIdleListener(new GoogleMap.OnCameraIdleListener() {
                    @Override
                    public void onCameraIdle() {
                        CameraPosition cameraPosition = map.getCameraPosition();
                        location = cameraPosition.target;
                        pin.setPosition(location);

                        if (geocoderAddressTask != null) {
                            geocoderAddressTask.cancel(true);
                        }
                        geocoderAddressTask = new GeocoderAddressTask(LocationFragment.this, fromPlaceSelected);
                        geocoderAddressTask.execute(location);
                        fromPlaceSelected = false;

                        if (autocompleteFragment != null) {
                            //autocompleteFragment.setBoundsBias(LatLngBounds.builder().include(location).build());
                            autocompleteFragment.setBoundsBias(new LatLngBounds(
                                    new LatLng(location.latitude - LOCATION_SEARCH_RADIUS, location.longitude - LOCATION_SEARCH_RADIUS),
                                    new LatLng(location.latitude + LOCATION_SEARCH_RADIUS, location.longitude + LOCATION_SEARCH_RADIUS)
                            ));
                        }
                    }
                });

                googleMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
                    @Override
                    public void onMapClick(final LatLng latLng) {
                        location = latLng;

                        CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLng(latLng);
                        map.moveCamera(cameraUpdate);

                        pin.setPosition(location);

                        hideKeyboard();
                    }
                });

                map = googleMap;
            }
        });
    }

    private void initializePlaces() {

        if (autocompleteFragment == null) {
            autocompleteFragment = new SupportPlaceAutocompleteFragment();
        }
        FragmentManager fragmentManager = getChildFragmentManager();
        fragmentManager.beginTransaction().replace(R.id.entourage_location_places, autocompleteFragment).commit();

//        AutocompleteFilter typeFilter = new AutocompleteFilter.Builder()
//                .setTypeFilter(AutocompleteFilter.TYPE_FILTER_ESTABLISHMENT)
//                .build();
//        autocompleteFragment.setFilter(typeFilter);

        if (originalLocation != null) {
            //autocompleteFragment.setBoundsBias(LatLngBounds.builder().include(originalLocation).build());
            autocompleteFragment.setBoundsBias(new LatLngBounds(
                    new LatLng(originalLocation.latitude - LOCATION_SEARCH_RADIUS, originalLocation.longitude - LOCATION_SEARCH_RADIUS),
                    new LatLng(originalLocation.latitude + LOCATION_SEARCH_RADIUS, originalLocation.longitude + LOCATION_SEARCH_RADIUS)
            ));
        } else {
            autocompleteFragment.setBoundsBias(new LatLngBounds(new LatLng(42, -5), new LatLng(51, 9)));
        }

        autocompleteFragment.setOnPlaceSelectedListener(new PlaceSelectionListener() {
            @Override
            public void onPlaceSelected(Place place) {
                if (place.getLatLng() != null) {
                    if (map != null) {
                        fromPlaceSelected = true;
                        CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLng(place.getLatLng());
                        map.moveCamera(cameraUpdate);
                        pin.setPosition(place.getLatLng());
                    }
                    addressTextView.setText(place.getAddress());
                }
                selectedPlace = place;
            }

            @Override
            public void onError(Status status) {
                if (getActivity() != null) {
                    Toast.makeText(getActivity(), R.string.entourage_location_address_not_found, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void initializeTimer() {
        handler = new Handler(Looper.getMainLooper());
        timerTask = new TimerTask() {
            @Override
            public void run() {
                if (autocompleteFragment == null) return;
                View autocompleteView = autocompleteFragment.getView();
                if (autocompleteView == null) return;
                TextView autocompleteEditText = autocompleteView.findViewById(com.google.android.gms.location.places.R.id.place_autocomplete_search_input);
                if (autocompleteEditText == null || TextUtils.isEmpty(autocompleteEditText.getText())) return;
                View autocompleteSearchView = autocompleteView.findViewById(com.google.android.gms.location.places.R.id.place_autocomplete_search_button);
                if (autocompleteSearchView != null) {
                    autocompleteSearchView.performClick();
                }
            }
        };
    }

    private void cancelTimer() {
        handler.removeCallbacks(timerTask);
    }

    private void setAddress(String address, boolean fromPlaceSelected) {
        addressTextView.setText(address);
        if (useGooglePlacesOnly && autocompleteFragment != null && !fromPlaceSelected) {
            View autocompleteView = autocompleteFragment.getView();
            if (autocompleteView != null) {
                TextView autocompleteEditText = autocompleteView.findViewById(com.google.android.gms.location.places.R.id.place_autocomplete_search_input);
                if (autocompleteEditText != null) {
                    cancelTimer();
                    handler.postDelayed(timerTask, SEARCH_DELAY);
                    autocompleteEditText.setText(address);
                }
            }
        }
    }

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------

    private static class GeocoderAddressTask extends AsyncTask<LatLng, Void, String> {

        boolean fromPlaceSelected;
        WeakReference<LocationFragment> locationFragmentWeakReference;

        private GeocoderAddressTask(LocationFragment locationFragment, boolean fromPlaceSelected) {
            this.fromPlaceSelected = fromPlaceSelected;
            locationFragmentWeakReference = new WeakReference<>(locationFragment);
        }

        @Override
        protected String doInBackground(final LatLng... params) {
            try {
                if (locationFragmentWeakReference == null) return null;
                LocationFragment locationFragment = locationFragmentWeakReference.get();
                if (locationFragment == null || locationFragment.getActivity() == null) {
                    return null;
                }
                Geocoder geoCoder = new Geocoder(locationFragment.getActivity(), Locale.getDefault());
                LatLng location = params[0];
                List<Address> addresses = geoCoder.getFromLocation(location.latitude, location.longitude, 1);
                String addressLine = "";
                if (addresses != null && addresses.size() > 0) {
                    Address address = addresses.get(0);
                    if (address.getMaxAddressLineIndex() >= 0) {
                        addressLine = addresses.get(0).getAddressLine(0);
                    }
                }
                return addressLine;
            }
            catch (IOException ignored) {

            }
            return null;
        }

        @Override
        protected void onPostExecute(final String address) {
            if (address == null || locationFragmentWeakReference == null) return;
            LocationFragment locationFragment = locationFragmentWeakReference.get();
            if (locationFragment == null || locationFragment.geocoderAddressTask != this) return;
            locationFragment.setAddress(address, fromPlaceSelected);
            locationFragment.geocoderAddressTask = null;
        }
    }

    public interface OnFragmentInteractionListener {
        void onEntourageLocationChosen(LatLng location, String address, Place place);
    }
}

File: app/src/main/java/social/entourage/android/location/LocationFragment.kt
package social.entourage.android.location

import android.Manifest.permission
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.location.Geocoder
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.TextView
import android.widget.Toast
import com.google.android.gms.common.api.Status
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener
import com.google.android.gms.maps.GoogleMapOptions
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import com.google.android.libraries.places.compat.Place
import com.google.android.libraries.places.compat.ui.PlaceSelectionListener
import com.google.android.libraries.places.compat.ui.SupportPlaceAutocompleteFragment
import kotlinx.android.synthetic.main.fragment_entourage_location.*
import kotlinx.android.synthetic.main.layout_view_title.*
import social.entourage.android.EntourageSecuredActivity
import social.entourage.android.R
import social.entourage.android.api.tape.Events.OnLocationPermissionGranted
import social.entourage.android.base.EntourageDialogFragment
import social.entourage.android.tools.BusProvider
import timber.log.Timber
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.*

/**
 * Fragment to choose the location of an entourage
 * Activities that contain this fragment must implement the
 * [LocationFragment.OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [LocationFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class LocationFragment  : EntourageDialogFragment() {
    // ----------------------------------
    // Attributes
    // ----------------------------------
    private var originalLocation: LatLng? = null
    private var originalAddress: String? = null
    private var useGooglePlacesOnly = false
    private var location: LatLng? = null
    private var selectedPlace: Place? = null
    private var mListener: OnFragmentInteractionListener? = null
    private var mapFragment: SupportMapFragment? = null
    private var map: GoogleMap? = null
    private var autocompleteFragment: SupportPlaceAutocompleteFragment? = null
    private var geocoderAddressTask: GeocoderAddressTask? = null
    private var handler: Handler? = null
    private var timerTask: TimerTask? = null
    private var fromPlaceSelected = false
    private var pin: Marker? = null

    // ----------------------------------
    // Lifecycle
    // ----------------------------------
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            originalLocation = it.getParcelable(KEY_ENTOURAGE_LOCATION)
            originalLocation?.let { loc ->
                location = LatLng(loc.latitude, loc.longitude)
            }
            originalAddress = it.getString(KEY_ENTOURAGE_ADDRESS)
            useGooglePlacesOnly = it.getBoolean(KEY_USE_GOOGLE_PLACES_ONLY, false)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        // Inflate the layout for this fragment
        val toReturn = inflater.inflate(R.layout.fragment_entourage_location, container, false)
        toReturn?.viewTreeObserver?.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                toReturn.let {
                    it.viewTreeObserver.removeOnGlobalLayoutListener(this)
                    dialog?.window?.let { window ->
                        val layoutParams = window.attributes
                        layoutParams.height = it.height
                        window.attributes = layoutParams
                    }
                }
            }
        })
        return toReturn
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeView()
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
        cancelTimer()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == PERMISSIONS_REQUEST_LOCATION) {
            for (index in permissions.indices) {
                if (permissions[index].equals(permission.ACCESS_FINE_LOCATION, ignoreCase = true)) {
                    val isGranted = grantResults[index] == PackageManager.PERMISSION_GRANTED
                    BusProvider.instance.post(OnLocationPermissionGranted(isGranted))
                    try {
                        map?.isMyLocationEnabled = isGranted
                    } catch (ex: SecurityException) {
                        Timber.e(ex)
                    }
                }
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    // ----------------------------------
    // INTERFACE CALLBACKS
    // ----------------------------------
    fun onBackClicked() {
        dismiss()
    }

    fun onValidateClicked() {
        mListener?.onEntourageLocationChosen(location, entourage_location_address?.text.toString(), selectedPlace)
        cancelTimer()
        dismiss()
    }

    private fun onCurrentLocationClicked() {
        (activity as EntourageSecuredActivity?)?.let {
            if (LocationUtils.isLocationPermissionGranted()) {
                EntourageLocation.currentLocation?.let { currentLocation ->
                    map?.moveCamera(CameraUpdateFactory.newLatLng(LatLng(currentLocation.latitude, currentLocation.longitude)))
                }
            } else {
                requestPermissions(arrayOf(permission.ACCESS_FINE_LOCATION), PERMISSIONS_REQUEST_LOCATION)
            }
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------
    private fun initializeView() {
        title_close_button.setOnClickListener { onBackClicked() }
        title_action_button.setOnClickListener { onValidateClicked() }
        entourage_location_current_position.setOnClickListener { onCurrentLocationClicked() }
        // Initialize map
        initializeMap()
        initializePlaces()
        initializeTimer()

        // Initialize address
        originalAddress?.let { setAddress(it, false) }
    }

    @SuppressLint("MissingPermission")
    private fun initializeMap() {
        if (mapFragment == null) {
            val googleMapOptions = GoogleMapOptions()
            googleMapOptions.zOrderOnTop(true)
            mapFragment = SupportMapFragment.newInstance(googleMapOptions)
        }
        mapFragment?.let { frag ->
            childFragmentManager.beginTransaction().replace(R.id.entourage_location_map_layout, frag).commit()
            frag.getMapAsync { googleMap ->
                if (activity != null) {
                    if (LocationUtils.isLocationPermissionGranted()) {
                        googleMap.isMyLocationEnabled = true
                    }
                }
                googleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(
                        activity, R.raw.map_styles_json))
                googleMap.uiSettings.isMyLocationButtonEnabled = false
                googleMap.uiSettings.isMapToolbarEnabled = false
                pin = if (originalLocation != null) {
                    val cameraUpdate = CameraUpdateFactory.newLatLngZoom(originalLocation, 15f)
                    googleMap.moveCamera(cameraUpdate)
                    val markerOptions = MarkerOptions()
                            .position(originalLocation!!)
                            .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_pin_orange))
                    googleMap.addMarker(markerOptions)
                } else {
                    val markerOptions = MarkerOptions()
                            .position(googleMap.cameraPosition.target)
                            .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_pin_orange))
                    googleMap.addMarker(markerOptions)
                }
                googleMap.setOnCameraMoveStartedListener { i ->
                    if (i == OnCameraMoveStartedListener.REASON_GESTURE) {
                        cancelTimer()
                    }
                }
                googleMap.setOnCameraIdleListener {
                    googleMap.cameraPosition.target?.let { loc ->
                        location = loc
                        pin?.position = loc
                        //cancelling old ones if exists
                        geocoderAddressTask?.cancel(true)
                        //starts new task
                        GeocoderAddressTask(this@LocationFragment, fromPlaceSelected).let { task ->
                            geocoderAddressTask = task
                            task.execute(location)
                        }
                        fromPlaceSelected = false
                        autocompleteFragment?.setBoundsBias(LatLngBounds(
                                LatLng(loc.latitude - LOCATION_SEARCH_RADIUS, loc.longitude - LOCATION_SEARCH_RADIUS),
                                LatLng(loc.latitude + LOCATION_SEARCH_RADIUS, loc.longitude + LOCATION_SEARCH_RADIUS)
                        ))
                    }
                }
                googleMap.setOnMapClickListener { latLng ->
                    location = latLng
                    googleMap.moveCamera(CameraUpdateFactory.newLatLng(latLng))
                    pin?.position = latLng
                    hideKeyboard()
                }
                map = googleMap
            }
        }
    }

    private fun initializePlaces() {
        if (autocompleteFragment == null) {
            autocompleteFragment = SupportPlaceAutocompleteFragment()
        }
        autocompleteFragment?.let { frag ->
            childFragmentManager.beginTransaction().replace(R.id.entourage_location_places, frag).commit()

            originalLocation?.let { loc ->
                frag.setBoundsBias(LatLngBounds(
                        LatLng(loc.latitude - LOCATION_SEARCH_RADIUS, loc.longitude - LOCATION_SEARCH_RADIUS),
                        LatLng(loc.latitude + LOCATION_SEARCH_RADIUS, loc.longitude + LOCATION_SEARCH_RADIUS)
                ))
            } ?: run  {
                frag.setBoundsBias(LatLngBounds(LatLng(42.0, -5.0), LatLng(51.0, 9.0)))
            }
            frag.setOnPlaceSelectedListener(object : PlaceSelectionListener {
                override fun onPlaceSelected(place: Place) {
                    place.latLng?.let { lat ->
                        fromPlaceSelected = true
                        map?.moveCamera(CameraUpdateFactory.newLatLng(lat))
                        pin?.position = lat
                        entourage_location_address?.text = place.address
                    }
                    selectedPlace = place
                }

                override fun onError(status: Status) {
                    activity?.let { Toast.makeText(it, R.string.entourage_location_address_not_found, Toast.LENGTH_SHORT).show() }
                }
            })
        }
    }

    private fun initializeTimer() {
        handler = Handler(Looper.getMainLooper())
        timerTask = object : TimerTask() {
            override fun run() {
                val autocompleteView = autocompleteFragment?.view ?: return
                val autocompleteEditText = autocompleteView.findViewById<TextView>(com.google.android.libraries.places.R.id.places_autocomplete_search_input) ?: return
                if (autocompleteEditText.text.isBlank()) return
                val autocompleteSearchView = autocompleteView.findViewById<View>(com.google.android.libraries.places.R.id.places_autocomplete_search_button)
                autocompleteSearchView?.performClick()
            }
        }
    }

    private fun cancelTimer() {
        timerTask?.let {handler?.removeCallbacks(it)}
    }

    private fun setAddress(address: String, fromPlaceSelected: Boolean) {
        entourage_location_address?.text = address
        if (useGooglePlacesOnly && !fromPlaceSelected) {
            autocompleteFragment?.view?.findViewById<TextView>(com.google.android.libraries.places.R.id.places_autocomplete_search_input)?.let { autocompleteEditText ->
                cancelTimer()
                timerTask?.let { handler?.postDelayed(it, SEARCH_DELAY)}
                autocompleteEditText.text = address
            }
        }
    }

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------
    private class GeocoderAddressTask(locationFragment: LocationFragment, var fromPlaceSelected: Boolean) : AsyncTask<LatLng?, Void?, String?>() {
        var locationFragmentWeakReference: WeakReference<LocationFragment>? = WeakReference(locationFragment)
        override fun doInBackground(vararg params: LatLng?): String? {
            try {
                params[0]?.let { location ->
                    locationFragmentWeakReference?.get()?.activity?.let { activity ->
                        val addresses = Geocoder(activity, Locale.getDefault())
                                .getFromLocation(location.latitude, location.longitude, 1)
                        if (addresses.size > 0) {
                            addresses?.first()?.let { address ->
                                if (address.maxAddressLineIndex >= 0) {
                                    return address.getAddressLine(0)
                                }
                            }
                        }
                        return ""
                    }
                }
            } catch (ignored: IOException) {
            }
            return null
        }

        override fun onPostExecute(address: String?) {
            if (address == null) return
            locationFragmentWeakReference?.get()?.let { locationFragment ->
                if (locationFragment.geocoderAddressTask !== this) return
                locationFragment.setAddress(address, fromPlaceSelected)
                locationFragment.geocoderAddressTask = null
            }
        }
    }

    interface OnFragmentInteractionListener {
        fun onEntourageLocationChosen(location: LatLng?, address: String?, place: Place?)
    }

    companion object {
        // ----------------------------------
        // Constants
        // ----------------------------------
        const val TAG = "social.entourage.android.entourage.location"
        private const val KEY_ENTOURAGE_LOCATION = "social.entourage.android.KEY_ENTOURAGE_LOCATION"
        private const val KEY_ENTOURAGE_ADDRESS = "social.entourage.android.KEY_ENTOURAGE_ADDRESS"
        private const val KEY_USE_GOOGLE_PLACES_ONLY = "social.entourage.android.KEY_USE_GOOGLE_PLACES_ONLY"
        private const val LOCATION_MOVE_DELTA = 50f //meters
        private const val LOCATION_SEARCH_RADIUS = 0.18f // 20 kilometers in lat/long degrees
        private const val PERMISSIONS_REQUEST_LOCATION = 1
        private const val SEARCH_DELAY = 1000L

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param location The initial location.
         * @param address The initial address as string
         * @param listener The listener
         * @return A new instance of fragment LocationFragment.
         */
        fun newInstance(location: LatLng?, address: String?, listener: OnFragmentInteractionListener?): LocationFragment {
            val args = Bundle()
            args.putParcelable(KEY_ENTOURAGE_LOCATION, location)
            args.putString(KEY_ENTOURAGE_ADDRESS, address)
            val fragment = LocationFragment()
            fragment.mListener = listener
            fragment.arguments = args
            return fragment
        }

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param location The initial location.
         * @param address The initial address as string
         * @param useGooglePlacesOnly true if the fragment will use only Google Places to fetch an address
         * @param listener The listener
         * @return A new instance of fragment LocationFragment.
         */
        @JvmStatic
        fun newInstance(location: LatLng?, address: String?, useGooglePlacesOnly: Boolean, listener: OnFragmentInteractionListener?): LocationFragment {
            val args = Bundle()
            args.putParcelable(KEY_ENTOURAGE_LOCATION, location)
            args.putString(KEY_ENTOURAGE_ADDRESS, address)
            args.putBoolean(KEY_USE_GOOGLE_PLACES_ONLY, useGooglePlacesOnly)
            val fragment = LocationFragment()
            fragment.mListener = listener
            fragment.arguments = args
            return fragment
        }
    }
}